File: backend/src/main/java/com/autocare/inventory/service/InventoryService.java
package com.autocare.inventory.service;

import com.autocare.inventory.dto.InventoryMovementDTO;
import com.autocare.inventory.dto.PartRequestDTO;
import com.autocare.inventory.dto.PartResponseDTO;
import com.autocare.inventory.entity.InventoryMovement;
import com.autocare.inventory.entity.Part;
import com.autocare.inventory.repository.InventoryMovementRepository;
import com.autocare.inventory.repository.PartRepository;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class InventoryService {

    private final PartRepository partRepository;
    private final InventoryMovementRepository movementRepository;

    @Transactional
    public PartResponseDTO createPart(PartRequestDTO request) {
        if (partRepository.findByCode(request.getCode()).isPresent()) {
            throw new BusinessException("Já existe uma peça com este código: " + request.getCode());
        }

        Part part = Part.builder()
                .name(request.getName())
                .code(request.getCode())
                .manufacturer(request.getManufacturer())
                .purchasePrice(request.getPurchasePrice())
                .salePrice(request.getSalePrice())
                .stockQuantity(request.getStockQuantity())
                .minimumStock(request.getMinimumStock() != null ? request.getMinimumStock() : 5)
                .isActive(true)
                .build();

        part = partRepository.save(part);

        // Registrar movimentação inicial
        InventoryMovement movement = InventoryMovement.createEntry(
                part,
                request.getStockQuantity(),
                "Cadastro inicial",
                "system"
        );
        movementRepository.save(movement);

        return toResponseDTO(part);
    }

    @Transactional
    public PartResponseDTO updatePart(String id, PartRequestDTO request) {
        Part part = findPartById(id);

        if (!part.getCode().equals(request.getCode()) &&
                partRepository.findByCode(request.getCode()).isPresent()) {
            throw new BusinessException("Já existe uma peça com este código: " + request.getCode());
        }

        part.setName(request.getName());
        part.setCode(request.getCode());
        part.setManufacturer(request.getManufacturer());
        part.setPurchasePrice(request.getPurchasePrice());
        part.setSalePrice(request.getSalePrice());
        part.setMinimumStock(request.getMinimumStock() != null ? request.getMinimumStock() : 5);

        part = partRepository.save(part);
        return toResponseDTO(part);
    }

    @Transactional
    public void deletePart(String id) {
        Part part = findPartById(id);
        part.setActive(false);
        partRepository.save(part);
    }

    @Transactional
    public void addStock(String partId, Integer quantity, String reason, String userId) {
        Part part = findPartById(partId);

        int previousQuantity = part.getStockQuantity();
        part.addStock(quantity);
        part = partRepository.save(part);

        InventoryMovement movement = InventoryMovement.createEntry(
                part, quantity, reason, userId
        );
        movementRepository.save(movement);
    }

    @Transactional
    public void removeStock(String partId, Integer quantity, String reason, String userId) {
        Part part = findPartById(partId);

        int previousQuantity = part.getStockQuantity();
        part.removeStock(quantity);
        part = partRepository.save(part);

        InventoryMovement movement = InventoryMovement.createExit(
                part, quantity, reason, userId
        );
        movementRepository.save(movement);
    }

    public PartResponseDTO findPartByIdResponse(String id) {
        return toResponseDTO(findPartById(id));
    }

    public List<PartResponseDTO> findAllParts() {
        return partRepository.findByIsActiveTrue()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<PartResponseDTO> findLowStockParts() {
        return partRepository.findLowStockParts()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<InventoryMovementDTO> getPartMovements(String partId) {
        findPartById(partId); // Verifica se a peça existe
        return movementRepository.findByPartIdOrderByCreatedAtDesc(partId)
                .stream()
                .map(this::toMovementDTO)
                .collect(Collectors.toList());
    }

    public Part findPartById(String id) {
        return partRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Peça não encontrada com ID: " + id));
    }

    private PartResponseDTO toResponseDTO(Part part) {
        return PartResponseDTO.builder()
                .id(part.getId())
                .name(part.getName())
                .code(part.getCode())
                .manufacturer(part.getManufacturer())
                .purchasePrice(part.getPurchasePrice())
                .salePrice(part.getSalePrice())
                .stockQuantity(part.getStockQuantity())
                .minimumStock(part.getMinimumStock())
                .isLowStock(part.isLowStock())
                .createdAt(part.getCreatedAt())
                .isActive(part.isActive())
                .build();
    }

    private InventoryMovementDTO toMovementDTO(InventoryMovement movement) {
        return InventoryMovementDTO.builder()
                .id(movement.getId())
                .partId(movement.getPart().getId())
                .partName(movement.getPart().getName())
                .type(movement.getType())
                .quantity(movement.getQuantity())
                .previousQuantity(movement.getPreviousQuantity())
                .currentQuantity(movement.getCurrentQuantity())
                .reason(movement.getReason())
                .referenceId(movement.getReferenceId())
                .userId(movement.getUserId())
                .createdAt(movement.getCreatedAt())
                .build();
    }
}

File: backend/src/inventory/inventory.service.ts
import { Injectable } from "@nestjs/common";
import { Prisma, MovementType } from "@prisma/client";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, InsufficientStockException, OptimisticLockException, ResourceNotFoundException } from "../common/exceptions";
import { PartRequestDto } from "./dto/part.dto";

@Injectable()
export class InventoryService {
  constructor(private prisma: PrismaService) {}

  async createPart(dto: PartRequestDto) {
    const existing = await this.prisma.part.findUnique({ where: { code: dto.code } });
    if (existing) {
      throw new BusinessException(`Já existe uma peça com este código: ${dto.code}`);
    }

    const part = await this.prisma.$transaction(async (tx) => {
      const created = await tx.part.create({
        data: {
          name: dto.name,
          code: dto.code,
          manufacturer: dto.manufacturer,
          purchasePrice: dto.purchasePrice,
          salePrice: dto.salePrice,
          stockQuantity: dto.stockQuantity,
          minimumStock: dto.minimumStock ?? 5,
          isActive: true,
        },
      });
      await tx.inventoryMovement.create({
        data: {
          partId: created.id,
          type: MovementType.ENTRADA,
          quantity: dto.stockQuantity,
          previousQuantity: 0,
          currentQuantity: dto.stockQuantity,
          reason: "Cadastro inicial",
          userId: "system",
        },
      });
      return created;
    });

    return this.toResponse(part);
  }

  async updatePart(id: string, dto: PartRequestDto) {
    const part = await this.findPartById(id);

    if (part.code !== dto.code) {
      const existing = await this.prisma.part.findUnique({ where: { code: dto.code } });
      if (existing) {
        throw new BusinessException(`Já existe uma peça com este código: ${dto.code}`);
      }
    }

    const updated = await this.prisma.part.update({
      where: { id },
      data: {
        name: dto.name,
        code: dto.code,
        manufacturer: dto.manufacturer,
        purchasePrice: dto.purchasePrice,
        salePrice: dto.salePrice,
        minimumStock: dto.minimumStock ?? 5,
      },
    });
    return this.toResponse(updated);
  }

  async deletePart(id: string) {
    await this.findPartById(id);
    await this.prisma.part.update({ where: { id }, data: { isActive: false } });
  }

  async addStock(partId: string, quantity: number, reason: string, userId: string) {
    if (quantity <= 0) {
      throw new BusinessException("Quantidade deve ser positiva");
    }
    await this.applyMovement(partId, quantity, reason, userId, MovementType.ENTRADA);
  }

  async removeStock(partId: string, quantity: number, reason: string, userId: string) {
    if (quantity <= 0) {
      throw new BusinessException("Quantidade deve ser positiva");
    }
    await this.applyMovement(partId, -quantity, reason, userId, MovementType.SAIDA);
  }

  // Lock otimista: lê a versão atual e condiciona o UPDATE a ela (where: { id, version }).
  // Se 0 linhas forem afetadas, outra transação já alterou a peça no meio do caminho —
  // equivalente ao OptimisticLockException que o Hibernate lançava no backend original
  // graças ao @Version em Part.java.
  private async applyMovement(
    partId: string,
    delta: number,
    reason: string,
    userId: string,
    type: MovementType,
  ) {
    const part = await this.findPartById(partId);
    const newQuantity = part.stockQuantity + delta;

    if (newQuantity < 0) {
      throw new InsufficientStockException(`Estoque insuficiente. Disponível: ${part.stockQuantity}`);
    }

    await this.prisma.$transaction(async (tx) => {
      const result = await tx.part.updateMany({
        where: { id: partId, version: part.version },
        data: { stockQuantity: newQuantity, version: { increment: 1 } },
      });
      if (result.count === 0) {
        throw new OptimisticLockException();
      }
      await tx.inventoryMovement.create({
        data: {
          partId,
          type,
          quantity: Math.abs(delta),
          previousQuantity: part.stockQuantity,
          currentQuantity: newQuantity,
          reason,
          userId,
        },
      });
    });
  }

  async getPartMovements(partId: string) {
    await this.findPartById(partId);
    const movements = await this.prisma.inventoryMovement.findMany({
      where: { partId },
      include: { part: true },
      orderBy: { createdAt: "desc" },
    });
    return movements.map((m) => this.toMovementResponse(m));
  }

  async findPartByIdResponse(id: string) {
    return this.toResponse(await this.findPartById(id));
  }

  async findAllParts() {
    const parts = await this.prisma.part.findMany({
      where: { isActive: true },
      orderBy: { createdAt: "desc" },
    });
    return parts.map((p) => this.toResponse(p));
  }

  async findLowStockParts() {
    const parts = await this.prisma.part.findMany({ where: { isActive: true } });
    return parts.filter((p) => p.stockQuantity <= p.minimumStock).map((p) => this.toResponse(p));
  }

  async findPartById(id: string) {
    const part = await this.prisma.part.findUnique({ where: { id } });
    if (!part) {
      throw new ResourceNotFoundException(`Peça não encontrada com ID: ${id}`);
    }
    return part;
  }

  private toResponse(part: any) {
    return {
      id: part.id,
      name: part.name,
      code: part.code,
      manufacturer: part.manufacturer,
      purchasePrice: part.purchasePrice,
      salePrice: part.salePrice,
      stockQuantity: part.stockQuantity,
      minimumStock: part.minimumStock,
      isLowStock: part.stockQuantity <= part.minimumStock,
      createdAt: part.createdAt,
      isActive: part.isActive,
    };
  }

  private toMovementResponse(movement: any) {
    return {
      id: movement.id,
      partId: movement.partId,
      partName: movement.part?.name,
      type: movement.type,
      quantity: movement.quantity,
      previousQuantity: movement.previousQuantity,
      currentQuantity: movement.currentQuantity,
      reason: movement.reason,
      referenceId: movement.referenceId,
      userId: movement.userId,
      createdAt: movement.createdAt,
    };
  }
}
